ENIGMAsystem/SHELL/Graphics_Systems/General/GSprimitives.cpp

```diff
@@ -1,17 +1,234 @@
-#include "GSprimitives.h"
-
-namespace enigma_user {
-
-unsigned draw_primitive_count(int kind, unsigned vertex_count) {
-  switch (kind) {
-    case pr_pointlist: return vertex_count;
-    case pr_linelist: return vertex_count / 2;
-    case pr_linestrip: if (vertex_count > 1) return vertex_count - 1; break;
-    case pr_trianglelist: return vertex_count / 3;
-    case pr_trianglestrip: if (vertex_count > 2) return vertex_count - 2; break;
-    case pr_trianglefan: if (vertex_count > 2) return vertex_count - 2; break;
-  }
-  return 0;
-}
-
-}
+/** Copyright (C) 2008-2013 Josh Ventura, Robert Colton
+*** Copyright (C) 2018 Robert Colton
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include "GSprimitives.h"
+#include "GSmodel.h"
+#include "GScolors.h"
+#include "GStextures.h"
+
+namespace {
+
+// the batching mode is initialized to the default here
+int draw_batch_mode = enigma_user::batch_flush_deferred;
+// the texture that was specified when the current primitive batch began
+int draw_batch_texture = -1;
+// lazy create the batch stream that we use for combining primitives
+int draw_get_batch_stream() {
+  static int draw_batch_stream = -1;
+  if (!enigma_user::d3d_model_exists(draw_batch_stream))
+    draw_batch_stream = enigma_user::d3d_model_create(enigma_user::model_stream);
+  return draw_batch_stream;
+}
```

Comment:
@JoshDreamland when you review this for me, please look over this particular spot very carefully because I was running into undefined behavior before I changed it to this (61ed4be743191ec2fa2473170c934f0b160da9b1).

Reply:
I did not recognize that this was static. Statics are initialized the first time the function is run. It should not behave differently now unless someone was destroying the resource this method returns. Is that the case? Also, is this method being called before `main()`?

Reply:
As far as I know, no, nobody was destroying it before. I don't believe it is being called before main. It would be called anywhere `draw_primitive_begin/_end` is called or `d3d_draw_*`. So it should only be getting called inside of a draw event.

Reply:
If you print every time you allocate a new buffer (every time you enter that `if`),  how often does it print?

Reply:
Apparently, just a single time.

```cpp
  if (!enigma_user::d3d_model_exists(draw_batch_stream)) {
    draw_batch_stream = enigma_user::d3d_model_create(enigma_user::model_stream);
    std::cerr << "new batch stream model created: " << darw_batch_stream << std::endl;
  }
```
![MSYS2 Log for Batch Stream Model Create](https://user-images.githubusercontent.com/3212801/44305402-86d4d980-a344-11e8-943e-c0c9cbc19935.png)

Reply:
```cpp
int draw_get_batch_stream() {
  static int draw_batch_stream = enigma_user::d3d_model_create(enigma_user::model_stream);
  return draw_batch_stream;
}
```
Yes, it goes back to the weird FPS dropping when focused issue again. Strange indeed.

Reply:
Print the IDs in both cases, please.

Reply:
@JoshDreamland alright, this is mildly depressing. If, after undoing the above shit, I enable link-time optimization for run mode? It results in the same FPS drop to 150fps in Direct3D9:
```cpp
	override CXXFLAGS += -Wall -s -O3 -fno-rtti -fno-exceptions -flto
	override CFLAGS += -Wall -s -O3 -flto -fno-exceptions
	override LDFLAGS += -flto -s -O3
```

We're getting bit somewhere, and I don't like it.

Reply:
I can reproduce that on master too all the way back 2016, I spent a whole day trying to find it. I am going to just put it aside for now and hope that you'll do me a solid some weekend and run Valgrind against master.